ejb/WFLY-10057-thread-pool.adoc

```diff
@@ -0,0 +1,165 @@
+
+
+|===
+|New Requirement |Existing Behavior of EJB3 Subsystem thread-pool
+
+|Ability to configure core-threads and max-threads independently
+|max-threads is configurable, but core-threads is not configurable and always equals to max-threads
+
+|Ability to reuse available threads as much as possible, without unnecessary creation of new threads
+|Upon new request, new threads are created up to the limit of max-threads, even though idle threads are available
+
+|Ability to timeout idle non-core threads after keepalive-timeout
+|Idle threads are not timed out, and keepalive-timeout is ignored
+
+|Ability to create and use non-core threads after core threads are saturated
+|Income tasks are queued after core threads are used up
+|===
+
+
+=== New elements in wildfly-ejb3_6_0.xsd schema
+
+Need to add a new sub-element under thread-pool element:
+
+[source]
+<xs:element name="core-threads" type="threads:countType"/>
+
+=== Expected CLI Operations and resources
+
+[source]
+/subsystem=ejb3/thread-pool=default:read-resource
+{
+"outcome" => "success",
+"result" => {
+  "core-threads" => 2,
+  "keepalive-time" => {
+    "time" => 1000L,
+    "unit" => "SECONDS"
+  },
+  "max-threads" => 10,
+  "name" => "default",
+  "thread-factory" => undefined
+}
+}
+
+[source]
+
+/subsystem=ejb3/thread-pool=default:write-attribute(name=core-threads, value=3)
+{"outcome" => "success"}
+
+[source]
+
+/subsystem=ejb3/thread-pool=default:read-attribute(name=core-threads)
+{
+"outcome" => "success",
+"result" => 3
+}
+
+[source]
+
+/subsystem=ejb3/thread-pool=default:write-attribute(name=max-threads, value=9)
+{"outcome" => "success"}
+
+[source]
+
+/subsystem=ejb3/thread-pool=default:read-attribute(name=max-threads)
+{
+"outcome" => "success",
+"result" => 9
+}
+
+=== New attribute in ejb3 subsystem resource model
+
+Need to add a new attribute to ejb3 subsystem resource model to represent core-threads configuration, as a peer
+to the existing max-threads attribute.
+
+=== Nice-to-Have Requirements
+
+* ability to switch between the existing thread-pool and the new thread-pool backed by `EnhancedQueueExecutor`
+
+=== Non-Requirements
+
+* improvement of thread-pool configuration in subsystems other than ejb3
+* ability to configure ejb3 thread-pool queue size
+* ability to allow ejb3 thread-pool core threads to timeout
+* improvement to thread-factory configuration in ejb3 subsystem thread-pool
+* configuration in ejb3 subsystem of advanced features in `EnhancedQueueExecutor`, e.g., growth resistance factor.
+
+
+== Implementation Plan
+////
+Delete if not needed. The intent is if you have a complex feature which can 
+not be delivered all in one go to suggest the strategy. If your feature falls 
+into this category, please mention the Release Coordinators on the pull 
+request so they are aware.
+////
+
+* switch ejb3 subsystem thread-pool to the thread-pool backed by `org.jboss.threads.EnhancedQueueExecutor`
```

Comment:
Will this be able to support all the metrics currently exposed by UnboundedQueueThreadPoolMetricsHandler?

```
[standalone@localhost:9990 /] /subsystem=ejb3/thread-pool=default:read-resource(include-runtime=true)
{
    "outcome" => "success",
    "result" => {
        "active-count" => 0,
        "completed-task-count" => 0L,
        "current-thread-count" => 0,
        "keepalive-time" => {
            "time" => 100L,
            "unit" => "MILLISECONDS"
        },
        "largest-thread-count" => 0,
        "max-threads" => 10,
        "name" => "default",
        "queue-size" => 0,
        "rejected-count" => 0,
        "task-count" => 0L,
        "thread-factory" => undefined
    }
}
```

Reply:
Yes, I've added a sample CLI output above (around line 121) that shows the same metrics.  It's taken after some custom config and test runs hence the non-default values.